File: src/parser/cursor.ts
import * as fs from 'fs';
import * as path from 'path';
import Database from 'better-sqlite3';
import { HistoryParser, Message, Session } from './types';

/** Shape of each tab entry stored in the Cursor DB. */
interface CursorTab {
  tabId?: string;
  chatTitle?: string;
  lastSendTime?: number;
  bubbles?: CursorBubble[];
}

interface CursorBubble {
  type?: string;   // "user" | "ai"
  text?: string;
  rawText?: string;
}

/**
 * Parses Cursor (and Windsurf) workspace storage.
 *
 * Cursor stores chat history in SQLite databases inside:
 *   ~/.config/Cursor/User/workspaceStorage/<hash>/state.vscdb
 *
 * The relevant rows in ItemTable have keys prefixed with
 * "interactive.sessions" or "workbench.panel.aichat".
 *
 * For workspaces that only leave JSON files (e.g. Windsurf), the parser also
 * reads any *.json file that contains a top-level "tabs" or "messages" array.
 */
export class CursorParser implements HistoryParser {
  canHandle(filePath: string): boolean {
    const base = path.basename(filePath);
    if (base === 'state.vscdb') return true;
    if (
      base.endsWith('.json') &&
      (filePath.includes('Cursor') || filePath.includes('Windsurf'))
    ) {
      return true;
    }
    return false;
  }

  async parse(filePath: string): Promise<Session[]> {
    if (filePath.endsWith('.vscdb')) {
      return parseCursorDB(filePath);
    }
    return parseCursorJSON(filePath);
  }
}

// ---- SQLite path -------------------------------------------------------

function parseCursorDB(filePath: string): Session[] {
  let db: Database.Database;
  try {
    db = new Database(filePath, { readonly: true });
  } catch {
    throw new Error(`cursor parser: opening ${filePath}`);
  }

  let rows: Array<{ value: string }>;
  try {
    rows = db
      .prepare(
        `SELECT value FROM ItemTable
         WHERE key LIKE 'interactive.sessions%'
            OR key LIKE 'workbench.panel.aichat%'
            OR key LIKE 'aiService.prompts%'`
      )
      .all() as Array<{ value: string }>;
  } catch {
    db.close();
    throw new Error(`cursor parser: querying ${filePath}`);
  }
  db.close();

  const sessions: Session[] = [];
  for (const row of rows) {
    try {
      const s = decodeCursorValue(row.value, filePath);
      sessions.push(...s);
    } catch {
      // Skip unrecognised shapes
    }
  }
  return sessions;
}

/** Tries several known JSON shapes from Cursor's DB. */
function decodeCursorValue(raw: string, dbPath: string): Session[] {
  // Shape 1: top-level array of tab objects.
  try {
    const tabs = JSON.parse(raw) as CursorTab[];
    if (Array.isArray(tabs) && tabs.length > 0) {
      return tabsToSessions(tabs, dbPath);
    }
  } catch {
    // fall through
  }

  // Shape 2: object with a "tabs" field.
  try {
    const wrapper = JSON.parse(raw) as { tabs?: CursorTab[] };
    if (wrapper.tabs && Array.isArray(wrapper.tabs) && wrapper.tabs.length > 0) {
      return tabsToSessions(wrapper.tabs, dbPath);
    }
  } catch {
    // fall through
  }

  throw new Error('unrecognised cursor value shape');
}

function tabsToSessions(tabs: CursorTab[], dbPath: string): Session[] {
  const project = path.dirname(path.dirname(dbPath));
  const sessions: Session[] = [];

  for (const tab of tabs) {
    const msgs: Message[] = [];
    for (const b of tab.bubbles ?? []) {
      let content = b.text ?? '';
      if (!content) content = b.rawText ?? '';
      content = content.trim();
      if (!content) continue;

      const role = b.type === 'ai' ? 'assistant' : 'user';
      msgs.push({ role, content });
    }
    if (msgs.length === 0) continue;

    let timestamp = new Date();
    if (tab.lastSendTime && tab.lastSendTime > 0) {
      // Cursor stores milliseconds since epoch.
      timestamp = new Date(tab.lastSendTime);
    }

    sessions.push({ tool: 'Cursor', project, timestamp, messages: msgs });
  }
  return sessions;
}

// ---- JSON path ---------------------------------------------------------

function parseCursorJSON(filePath: string): Session[] {
  const data = fs.readFileSync(filePath, 'utf8');
  let parsed: unknown;
  try {
    parsed = JSON.parse(data);
  } catch {
    throw new Error(`cursor parser: invalid JSON in ${filePath}`);
  }

  // Try decoding as an array of tab objects.
  if (Array.isArray(parsed) && parsed.length > 0) {
    const project = path.dirname(path.dirname(filePath));
    const fakePath = path.join(project, 'state.vscdb');
    return tabsToSessions(parsed as CursorTab[], fakePath);
  }

  // Try decoding as an object with a "messages" array (simpler shape).
  if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
    const obj = parsed as { messages?: Array<{ role?: string; content?: string }> };
    if (obj.messages && Array.isArray(obj.messages) && obj.messages.length > 0) {
      const msgs: Message[] = [];
      for (const m of obj.messages) {
        if (!m.content) continue;
        msgs.push({ role: m.role ?? 'user', content: m.content });
      }
      const project = path.dirname(path.dirname(filePath));
      return [{ tool: 'Cursor', project, timestamp: new Date(), messages: msgs }];
    }
  }

  throw new Error(`cursor parser: unrecognised JSON shape in ${filePath}`);
}

File: internal/parser/cursor.go
package parser

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	_ "modernc.org/sqlite" // pure-Go SQLite driver (no cgo required)
)

// CursorParser parses Cursor (and Windsurf) workspace storage.
//
// Cursor stores chat history in SQLite databases inside:
//
//	~/.config/Cursor/User/workspaceStorage/<hash>/
//	    state.vscdb          ← main key/value store
//
// The relevant rows in the ItemTable have keys prefixed with
// "interactive.sessions" or "workbench.panel.aichat".  The values are JSON
// blobs containing arrays of chat messages.
//
// For workspaces that only leave JSON files (e.g. Windsurf), the parser also
// reads any *.json file that contains a top-level "tabs" or "messages" array.
type CursorParser struct{}

// CanHandle returns true for state.vscdb SQLite files or chat JSON files
// inside a Cursor/Windsurf workspace storage directory.
func (p *CursorParser) CanHandle(path string) bool {
	base := filepath.Base(path)
	slash := filepath.ToSlash(path)
	if base == "state.vscdb" {
		return true
	}
	if strings.HasSuffix(base, ".json") &&
		(strings.Contains(slash, "Cursor") || strings.Contains(slash, "Windsurf")) {
		return true
	}
	return false
}

// Parse extracts chat sessions from a Cursor workspace storage file.
func (p *CursorParser) Parse(path string) ([]Session, error) {
	if strings.HasSuffix(path, ".vscdb") {
		return parseCursorDB(path)
	}
	return parseCursorJSON(path)
}

// ---- SQLite path -------------------------------------------------------

// cursorTabJSON is the shape of each tab entry stored in the DB.
type cursorTabJSON struct {
	ID        string         `json:"tabId"`
	Title     string         `json:"chatTitle"`
	Timestamp int64          `json:"lastSendTime"`
	Bubbles   []cursorBubble `json:"bubbles"`
}

type cursorBubble struct {
	Type    string `json:"type"` // "user" | "ai"
	Text    string `json:"text"`
	RawText string `json:"rawText"`
}

func parseCursorDB(path string) ([]Session, error) {
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, fmt.Errorf("cursor parser: opening %s: %w", path, err)
	}
	defer db.Close()

	// Cursor stores its chat state in ItemTable with varying key prefixes.
	rows, err := db.Query(
		`SELECT value FROM ItemTable
         WHERE key LIKE 'interactive.sessions%'
            OR key LIKE 'workbench.panel.aichat%'
            OR key LIKE 'aiService.prompts%'`)
	if err != nil {
		return nil, fmt.Errorf("cursor parser: querying %s: %w", path, err)
	}
	defer rows.Close()

	var sessions []Session
	for rows.Next() {
		var raw string
		if err := rows.Scan(&raw); err != nil {
			continue
		}
		s, err := decodeCursorValue(raw, path)
		if err != nil {
			continue
		}
		sessions = append(sessions, s...)
	}
	return sessions, nil
}

// decodeCursorValue tries several known JSON shapes from Cursor's DB.
func decodeCursorValue(raw, dbPath string) ([]Session, error) {
	// Shape 1: top-level array of tab objects.
	var tabs []cursorTabJSON
	if err := json.Unmarshal([]byte(raw), &tabs); err == nil && len(tabs) > 0 {
		return tabsToSessions(tabs, dbPath), nil
	}
	// Shape 2: object with a "tabs" field.
	var wrapper struct {
		Tabs []cursorTabJSON `json:"tabs"`
	}
	if err := json.Unmarshal([]byte(raw), &wrapper); err == nil && len(wrapper.Tabs) > 0 {
		return tabsToSessions(wrapper.Tabs, dbPath), nil
	}
	return nil, fmt.Errorf("unrecognised cursor value shape")
}

func tabsToSessions(tabs []cursorTabJSON, dbPath string) []Session {
	project := filepath.Dir(filepath.Dir(dbPath))
	var sessions []Session
	for _, tab := range tabs {
		var msgs []Message
		for _, b := range tab.Bubbles {
			content := b.Text
			if content == "" {
				content = b.RawText
			}
			content = strings.TrimSpace(content)
			if content == "" {
				continue
			}
			role := "user"
			if b.Type == "ai" {
				role = "assistant"
			}
			msgs = append(msgs, Message{Role: role, Content: content})
		}
		if len(msgs) == 0 {
			continue
		}
		ts := time.Now().UTC()
		if tab.Timestamp > 0 {
			// Cursor stores milliseconds since epoch.
			ts = time.UnixMilli(tab.Timestamp).UTC()
		}
		sessions = append(sessions, Session{
			Tool:      "Cursor",
			Project:   project,
			Timestamp: ts,
			Messages:  msgs,
		})
	}
	return sessions
}

// ---- JSON path ---------------------------------------------------------

// parseCursorJSON handles plain JSON files that some Cursor/Windsurf versions
// write out.
func parseCursorJSON(path string) ([]Session, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("cursor parser: reading %s: %w", path, err)
	}

	// Try decoding as an array of tab objects.
	var tabs []cursorTabJSON
	if err := json.Unmarshal(data, &tabs); err == nil && len(tabs) > 0 {
		project := filepath.Dir(filepath.Dir(path))
		return tabsToSessions(tabs, filepath.Join(project, "state.vscdb")), nil
	}

	// Try decoding as an object with a "messages" array (simpler shape).
	var simple struct {
		Messages []struct {
			Role    string `json:"role"`
			Content string `json:"content"`
		} `json:"messages"`
	}
	if err := json.Unmarshal(data, &simple); err == nil && len(simple.Messages) > 0 {
		var msgs []Message
		for _, m := range simple.Messages {
			if m.Content == "" {
				continue
			}
			msgs = append(msgs, Message{Role: m.Role, Content: m.Content})
		}
		project := filepath.Dir(filepath.Dir(path))
		return []Session{{
			Tool:      "Cursor",
			Project:   project,
			Timestamp: time.Now().UTC(),
			Messages:  msgs,
		}}, nil
	}

	return nil, fmt.Errorf("cursor parser: unrecognised JSON shape in %s", path)
}
